architecture/2024-05-radius-on-dapr.md

```diff
@@ -0,0 +1,324 @@
+
+### Goals
+
+- Adopt Dapr for our data storage and messaging needs. 
+  - Keep `rad init` and `rad install` simple by including Dapr and a default choice of components as part of our Helm chart.
+  - Keep the Helm chart flexible by documenting how to BYO components.
+  - Provide production-guidance for Radius based on Dapr's existing guidance.
+- Remove our existing messaging and storage implementations in favor of Dapr.
+
+### Non goals
+
+- (out of scope) User-facing behavior changes other than installation/dependencies.
+- (out of scope/follow-up) Adoption of Dapr workflows.
+  - While this is highly valuable for us, this 
+
+### User scenarios (optional)
+
+#### Wendy tries Radius out
+
+Wendy wants to learn more about Radius, so she fires up a Kubernetes cluster and tries some of our tutorials. She runs `rad init` to install Radius, which includes Dapr and some components that Radius can use for its storage. This is easy for her because she didn't need to learn about Dapr to try Radius out.
+
+*This scenario is here to point out that the simplicity of the current user experience is maintained.*
+
+#### Wendy and Jake put Radius in production
+
+Wendy's been successful experimenting with Radius, and she wants to take it to production. She works with Jake to set up some cloud datastores for Radius to use, to ensure that data is persisted. Jake creates a DynamoDB table and the necessary SNS/SQS resources, following the Dapr documentation. Jake reads the Radius documentation, installs Dapr, and configures the `Component` objects in Kubernetes that are required. Lastly, Jake runs `helm install radius radius -n radius-system` to install Radius without its default component configuration.
+
+
+## User Experience (if applicable)
+
+**Sample Input:**
+
+*n/a*
+
+**rad init Sample Output:**
+
+Running `rad init`:
+
+Before 
+
+```txt
+✅ Installing Radius 0.32.0 on kind-kind
+✅ Create new environment default
+   - Kubernetes namespace: default
+   - Recipe pack: local-dev
+✅ Update local configuration
+```
+
+
+After 
+
+```txt
+✅ Installing Radius 0.32.0 on kind-kind
+✅ Installing Dapr 1.13.0 on kind-kind
+✅ Create new environment default
+   - Kubernetes namespace: default
+   - Recipe pack: local-dev
+✅ Update local configuration
+```
+
+
+**Sample Recipe Contract:**
+
+*n/a*
+
+### Development Experience
+
+*Adding Dapr as a dependency has the potential to impact our development experience. We can follow some existing patterns that Dapr users like to simplify this.*
+
+Wilford is a developer that's making some contributions to Radius for the first time. To get started, he reads the developer document at in [the repo](https://github.com/radius-project/radius/blob/main/docs/contributing/contributing-code/contributing-code-control-plane/running-controlplane-locally.md). He sees that installing Dapr is part of the prerequisites, so he follows the [getting started instructions](https://docs.dapr.io/getting-started/).
```

Comment:
Why is it a prerequisite? Is Radius going to depend on local **Dapr** installation?

Reply:
Yes. If Dapr is required, then it's required. Do you see this as a problem?

Reply:
If I don't know about **Dapr**, then to use **Radius** I will need to go and install **Dapr**. It feels like it is one additional step for the user. Not a big problem if we expect/assume users of Radius to also know about Dapr.

Another thought is that if I am installing a software and, to use it, it asks me to install another one. Basically, I need to install 2 software. 🤔

We can discuss further.

Reply:
Let's discuss this again next time we meet. It's the primary topic of the document.

We're going to add **some** kind of dependency to Radius, that's not up for debate.